cmd/influx/query.go

```diff
@@ -0,0 +1,138 @@
+		ArgsUsage:   "[query literal or '-' for stdin]",
+		Before:      middleware.WithBeforeFns(withCli(), withApi(true)),
+		Flags: append(
+			commonFlagsNoPrint,
+			&cli.GenericFlag{
+				Name:    "org-id",
+				Usage:   "The ID of the organization",
+				EnvVars: []string{"INFLUX_ORG_ID"},
+				Value:   &orgParams.OrgID,
+			},
+			&cli.StringFlag{
+				Name:        "org",
+				Usage:       "The name of the organization",
+				Aliases:     []string{"o"},
+				EnvVars:     []string{"INFLUX_ORG"},
+				Destination: &orgParams.OrgName,
+			},
+			&cli.StringFlag{
+				Name:    "file",
+				Usage:   "Path to Flux query file",
+				Aliases: []string{"f"},
+			},
+			&cli.BoolFlag{
+				Name:    "raw",
+				Usage:   "Display raw query results",
+				Aliases: []string{"r"},
+			},
+			&cli.StringSliceFlag{
+				Name:    "profilers",
+				Usage:   "Names of Flux profilers to enable",
+				Aliases: []string{"p"},
+			},
+		),
+		Action: func(ctx *cli.Context) error {
+			queryString, err := readQuery(ctx)
+			if err != nil {
+				return err
+			}
+			queryString = strings.TrimSpace(queryString)
+			if queryString == "" {
+				return errors.New("no query provided")
+			}
+
+			// The old CLI allowed specifying this either via repeated flags or
+			// via a single flag w/ a comma-separated value.
+			rawProfilers := ctx.StringSlice("profilers")
+			var profilers []string
+			for _, p := range rawProfilers {
+				profilers = append(profilers, strings.Split(p, ",")...)
+			}
+
+			params := query.Params{
+				OrgParams: orgParams,
+				Query:     queryString,
+				Profilers: profilers,
+			}
+
+			printer := query.RawResultPrinter
+			if !ctx.Bool("raw") {
+				return errors.New("--raw or -r must be specified for now")
+			}
+
+			client := query.Client{
+				CLI:           getCLI(ctx),
+				QueryApi:      getAPI(ctx).QueryApi,
+				ResultPrinter: printer,
+			}
+			return client.Query(ctx.Context, &params)
+		},
+	}
+}
+
+// readQuery reads a Flux query into memory from a file, args, or stdin based on CLI parameters.
+func readQuery(ctx *cli.Context) (string, error) {
+	nargs := ctx.NArg()
+	file := ctx.String("file")
+
+	if nargs > 1 {
+		return "", fmt.Errorf("at most 1 query string can be specified over the CLI, got %d", ctx.NArg())
+	}
+	if nargs == 1 && file != "" {
+		return "", errors.New("query can be specified via --file or over the CLI, not both")
+	}
+
+	readFile := func(path string) (string, error) {
+		queryBytes, err := ioutil.ReadFile(path)
+		if err != nil {
+			return "", fmt.Errorf("failed to read query from %q: %w", path, err)
+		}
+		return string(queryBytes), nil
+	}
+
+	readStdin := func() (string, error) {
+		queryBytes, err := ioutil.ReadAll(os.Stdin)
+		if err != nil {
+			return "", fmt.Errorf("failed to read query from stdin: %w", err)
+		}
+		return string(queryBytes), err
+	}
+
+	if file != "" {
+		return readFile(file)
+	}
+	if nargs == 0 {
+		return readStdin()
+	}
```

Comment:
I'm wondering if there's a way to combine some of the logic from line 106 through 127. Could you have a single function that takes an `io.Reader` (or even an `*os.File`) instead of the separate `readFile` and `readStdin` functions? Then if you've got a file, open it with `os.Open` first and pass it into the function.

Reply:
I originally made a single function that took an `io.Reader`, but it just pushed the redundancies around: each of the current `readFile` calls were replaced by something like:
```go
f, err := os.Open(path)
if err != nil {
  return err
}
defer f.Close()
query, err := readQuery(f, path)
if err != nil {
  return err
}
```
Making separate functions for read-path and read-stdin at least lets us use different `ioutil` conveniences and cut down the boilerplate.